test: fatal exception: Add code description and testcases

yingmingx opened the pull request:
1. Improve the test case of CPU exception.
2. Add equivalence classes and input partition testing
when give an integer reason code.

Signed-off-by: Ying ming <mingx.ying@intel.com>

ioannisg commented:
There are some riscv board failures afaics

yingmingx commented:
> There are some riscv board failures afaics

Yes, I need to find a way to trigger exceptions on the riscv platform.

nashif reviewed (changes requested):
you need to split this change into 2 commits:
1. doxygen changes
2. adding a new test..

yingmingx commented:
> you need to split this change into 2 commits:
> 
> 1. doxygen changes
> 2. adding a new test..

Hi, Anas. I have split this change into 2 commits.

LeiW000 commented on tests/arch/x86/static_idt/src/main.c:
```diff
@@ -150,11 +150,41 @@ void idt_spur_task(void *arg1, void *arg2, void *arg3)
+ * Test Procedure:
+ * -# Trigger with assembly instructions to see interrupt handler
+ * executes properly.
+ * -# Make an operands divide by zero to see exception handler
```
I guess that "operands" shoud be "operand".

enjiamai commented on tests/kernel/fatal/exception/src/main.c:
```diff
@@ -90,6 +90,35 @@ void entry_cpu_exception(void *p1, void *p2, void *p3)
+	 */
+	__asm__ volatile (".word 0x77777777");
+#elif defined(CONFIG_ARC)
+	__asm__ volatile ("swi");
```
This function looks like to test the scenario a divide zero exception happened, for x86, it is.  But the exceptions here for other architecture seems not intend to do this, could you please explain why you do this here? thanks!

enjiamai replied:
Ok, see the function name change, I got it.

enjiamai commented on tests/kernel/fatal/exception/src/main.c:
```diff
@@ -90,6 +90,35 @@ void entry_cpu_exception(void *p1, void *p2, void *p3)
 	rv = TC_FAIL;
 }
 
+void entry_cpu_exception_divide_zero(void *p1, void *p2, void *p3)
```
According to below question, would it be better to add some descriptions for this subroutine? thanks!

enjiamai reviewed (changes requested):
Have some questions, could you please help to clarify? thanks!

yingmingx commented:
> Have some questions, could you please help to clarify? thanks!

Hi, enjia. A note is added to explain why the test case is added and the test process.

LeiW000 commented:
@nashif Could you help to merge it? I think it is ready for merging.

yingmingx commented:
This PR has rebased.

enjiamai commented:
@yingmingx  I was thinking if it is need to remove the doxygen commit, just keep the code change, It might be easier for this to be merged.

Commit:
test: fatal exception: add test case

1. Improve the test case of CPU exception.
2. Add equivalence classes and input partition testing
when give an integer reason code.

Signed-off-by: Ying ming <mingx.ying@intel.com>

ioannisg commented on tests/kernel/fatal/exception/src/main.c:
```diff
@@ -90,6 +90,38 @@ void entry_cpu_exception(void *p1, void *p2, void *p3)
+	__asm__ volatile ("svc 0");
+#elif defined(CONFIG_CPU_CORTEX_R)
+	__asm__ volatile ("BKPT");
+#elif defined(CONFIG_ARM)
```
Could this be CPU_CORTEX_M, to make it explicit? That's because CONFIG_ARM is common to AARCH32 -R and -M

ioannisg replied:
Perhaps you can also trigger the SVC with the SVC instruction directly?

yingmingx replied:
> Perhaps you can also trigger the SVC with the SVC instruction directly?

Thank you for you advice. I tried to use SVC directly. But it failed on build qemu_cortex_a53,
/tmp/ccWs7opR.s:1185: Error: missing immediate expression at operand 1 -- `svc'

yingmingx replied:
> Could this be CPU_CORTEX_M, to make it explicit? That's because CONFIG_ARM is common to AARCH32 -R and -M

I distinguish CONFIG_CPU_CORTEX_R because the platform qemu_cortex_r5 will fail when using "swi 0", so I use another instruction to make fatal error insted of "swi 0".
When using CPU_CORTEX_M instead of CONFIG_ARM, building platform on mps2_an385 and mps2_521 will fail. The macro in  .config file show there is only CONFIG_ARM without CPU_CORTEX_M. So I think CONFIG_ARM is a more common macro to make the compiler build pass.

ioannisg commented on tests/kernel/fatal/exception/src/main.c:
```diff
@@ -138,6 +170,19 @@ void entry_arbitrary_reason(void *p1, void *p2, void *p3)
+
+	expected_reason = -2;
+
+	key = irq_lock();
```
What was the need for locking IRQs here?

yingmingx replied:
I copy this lock IRQ from other testcase in this file.
I think the author write this code here is because expected_reason is a global variable,  use irq_lock() here will explicitly explain that the variable would not be changed in the isr context.

ioannisg reviewed (commented):
Looks ok, but I would like to hear back from @yingmingx around the comments i left

enjiamai commented:
@yingmingx Please rebase it~

yingmingx commented:
Close this PR. The new one is #32769.
